// flocoto.integration/data.ts
import tns from "./catalog/tns";

namespace flocoto_data {


    export type AuthUserInfo = {
        user_uid: string,
        user_mail: string,
        user_name: string,
        user_surname: string,
        user_photo: string,
        _branch_uid: string,
        return_type: string,
        user_access: number,
        user_type: number,
        user_gender: string,
        user_status: number,
        user_fullname: string,
        user_lastchangepassword: string
    }

    export type AuthInfo = {
        result: true,
        auth_key: string,
        auth_block?: boolean,
        auth_application?: string,
        auth_uid: string,
        auth_date: string
        auth_status: number,
        auth_endlife: string,
        auth_viewport: string,
        auth_flocoto: string,
        auth_session: string,
    }

    export type AuthBranch = {
        branch_uid:string
        branch_menus:number[]
        branch_mainuser:string
        branch_mainworkspace:string
    }

    export type AuthGrants = {
        acesso_uid:string
        menu_id:number
        menu_menu_id:number
        menu_codigo:string
        menu_nivel:number
        menu_icon:string
        menu_nome:string
        menu_link:string
        menu_app:string
        menu_viewports:string[]
    };

    export type AuthRole = tns.PartialOf<"auth.cargo"|"auth.espaco"|"auth.trabalha", "cargo_"|"espaco_"|"trabalha_",
        "uid"
        |"nome"
        |"code"
        |"estado"
        |"inicio"
        |"fim"
        |"default"
        |"meta"
        |"cargo_uid"
        |"espaco_work"
        |"user_owner"
    >

    export interface Viewport {
        viewport_identifier:string
        viewport_name:string
        viewport_desc:string
        viewport_application:string
        viewport_from:string
        viewport_origin:string
        viewport_proxypath:string
        viewport_status:number
        viewport_visible:boolean
        viewport_session:string
    }
}

export = flocoto_data;
